### DISCUSSION TITLE
How to test `shared_module`?

### DISCUSSION THREAD

**Discussion by chriselrod** [mesonbuild/meson#11208] [Q&A] [answered]
I have a project that builds a shared module:
```meson
my_shared_module = shared_module(...)
```

I would like to test it via running an executable that supports loading modules
```cxx
FILE *fopt = popen("opt -load-pass-plugin=path/to/my_shared_module.so moreargs path/to/file.ll", "r")
```
and then compare the results with some expected results:
```cxx
FILE *ftxt = fopen("path/to/cmpfile.txt", "r")
```

However, I have two problems here:
1. I do not know how to declare `my_shared_module.so` as a dependency of the test. It is not a shared library, but a shared module loaded by the executable.
2. I do not know how to set the relative paths correctly so that this works on CI.

Or, is there a better approach?

**Comment by chriselrod** [comment]
Perhaps I can solve the path via defining a macro to be `meson.source_root()`?

**Comment by eli-schwartz** [comment, marked as answer]
> 1. I do not know how to declare `my_shared_module.so` as a dependency of the test. It is not a shared library, but a shared module loaded by the executable.

```meson
test('name', exe, ..., depends: [my_shared_module])
```

> 2. I do not know how to set the relative paths correctly so that this works on CI.

How are you running it? If you pass a built target as part of the `args:` kwarg, Meson will guarantee the path resolves correctly to the built object.

There's also `my_shared_module.full_path()` to yield a string. This may be necessary if it must be `--flag=path/to/module` instead of `--flag path/to/module`. But in that case you *need* `depends:` as meson cannot figure out dependencies automatically from a string.

**Reply by chriselrod** [reply]
Thanks,
```meson
test('name', exe, ..., depends: [my_shared_module])
```
works. Running `meson test` now causes `my_shared_module` to build, and it also recompiles when edited and rerunning `meson test`.

I'll try passing paths now.

> How are you running it?

I was simple running `meson test`.

> If you pass a built target as part of the args: kwarg, Meson will guarantee the path resolves correctly to the built object.

This will pass command line args?
I was currently using `gtest_main`, so didn't have a manually written main function.
But I don't really need `gtest` for this.

> There's also my_shared_module.full_path() to yield a string. This may be necessary if it must be --flag=path/to/module instead of --flag path/to/module. But in that case you need depends: as meson cannot figure out dependencies automatically from a string.

I need paths to:
1. the shared module, which is loaded by the executable.
2. a file that is passed to the executable as an argument.
3. an output file I can compare the executable's output to.

I'm going to be using `depends:`, of course, but I figure `my_shared_module.full_path()` is what I'd want for `1.`, and then `meson.source_root()` can be used for `2.` and `3.`.

**Reply by eli-schwartz** [reply]
For 2, you probably want to pass `files('foo.txt')` as an argument.

For 3, probably the same (although it wasn't clear from your wording whether it is a custom_target build output).

**Reply by chriselrod** [reply]
For reference, I am currently doing:
```meson
test(..., args: [turbo_module.full_path(), meson.source_root() / 'test' / 'examples'])
```
https://github.com/JuliaSIMD/LoopModels/blob/f726b90f3417aa7049aaa002f3b2e79357947ea6/meson.build#L34
and then using these to `popen` (which runs an executable, and returns a file handle to the `stdout`) and `fopen`:
```cxx
  std::string modulePath = argv[1];
  std::string examplesPath = argv[2];
  printf("modulePath: %s\n", modulePath.c_str());
  printf("examplesPath: %s\n", examplesPath.c_str());
  std::string fileRoot = examplesPath + std::string("/triangular_solve.");
  auto cmd =
    std::string(
      "opt -mcpu=skylake-avx512 --disable-output -load-pass-plugin=") +
    modulePath +
    std::string(" -passes='turbo-loop' -pass-remarks-analysis='turbo-loop' ") +
    fileRoot + std::string("ll 2>&1");
  printf("cmd: %s", cmd.c_str());
  std::string txtfile = fileRoot + std::string("txt");

  FILE *fpopt = popen(cmd.c_str(), "r");
  FILE *fptxt = fopen(txtfile.c_str(), "r");
```
https://github.com/JuliaSIMD/LoopModels/blob/f726b90f3417aa7049aaa002f3b2e79357947ea6/test/remarks_test.cpp#L9-L24

So an example `cmd` (this one taken from a CI log) that gets generated and run is:
```sh
opt -mcpu=skylake-avx512 --disable-output -load-pass-plugin=/Users/runner/work/LoopModels/LoopModels/builddir/libTurboLoop.dylib -passes='turbo-loop' -pass-remarks-analysis='turbo-loop' /Users/runner/work/LoopModels/LoopModels/test/examples/triangular_solve.ll 2>&1
```
This seems to work: tests pass locally and on CI.

But I'm open to a better way to do things.

To be clear, "2." and "3." are both simply files that are checked into the repo. They aren't produced by a build, but used for testing (I'll likely add more such files later, I just wanted to get an example working first for automatic regression checking).
This is why I passed the path for the directory containing these files as `meson.source_root() / 'test' / 'examples'`.

But, because I may change these files, perhaps I should also declare them as dependencies?
